[Woo POS] Products grid

samiuelson opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #11516 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR adds a very basic grid view showing products at the left side of the Cart screen. The product grid supports pagination.

Not included:
* Filtering simple products
* Loading indicators
* Item selection
* Unit tests

The above will come in the next PRs for the sake of brevity.

### Testing instructions
<!-- Step-by-step testing instructions. When necessary, break out individual scenarios that need testing, and consider including a checklist for the reviewer to go through. -->
Test if all the products are loading. At the current stage all products should be visible, not only simple products.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/woocommerce/woocommerce-android/assets/4527432/a6c597a9-7a45-44be-8be1-6105d8052fd5

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

samiuelson committed:
Update WooPosCartScreen

samiuelson committed:
Merge branch 'trunk' into 11516-woo-pos-core-home-screen-ui--product-grid-cart-top-app-bar

samiuelson committed:
Implement products data source

samiuelson committed:
Implement base product selector vm

samiuelson committed:
Display grid with product titles

samiuelson committed:
Address detekt's complaints

dangermattic commented:
<!--
  0 Errors
  3 Warnings: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          3 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>ProductSelectorViewModel</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>ProductsDataSourceImpl</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

samiuelson committed:
Implement infinite grid loading

backwardstruck commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/di/POSModule.kt:
```diff
@@ -0,0 +1,20 @@
+class POSModule {
+    @Provides
+    fun provideProductList(
+        handler: ProductListHandler
```
Seems like we should indeed use this as it will be faster than making a new class and so far there's no differences with POS requirements. I think we can swap it out later if the POS functionality requires changes.

malinajirka replied:
Some questions that come to mind which might be worth answering: How complex will it be to revert this decision later? What are the risks of affecting the store-management section of the app (eg. introducing new bugs, etc)? Does this decision make the store-management section more complex? Does it increase maintainability (eg. more testing is needed, code is more difficult to reason about, ...)? What are reasons for change to the ProductListHandler - is it dependant on the use-case (POS vs store-management) or is it a generic component?

Just to be clear, I'm not challenging whether we should re-use or not, since I honestly don't know for this case. I'm just trying to better understand the implications. In general, I'm all up for re-using where we can, but at the same time I don't think we should re-use just because the code can be invoked from two different sections of the app and works (now) -> the reason for change should imo be the same for all the sections in which the code is used or the code probably shouldn't be re-used.

samiuelson replied:
Thanks for raising these questions, @malinajirka. Let me share my POV.
> How complex will it be to revert this decision later?

My reasoning is we won't need to revert this decision. We may want to write a new implementation of the `ProductsDataSource` interface (I don't see a clear reason for doing it now though):
```kotlin
interface ProductsDataSource {
    val products: Flow<List<Product>>
    suspend fun loadProducts()
    suspend fun loadMore()
}
```
For now, employing the existing `ProductListHandler`'s code to implement the interface saves us time. I think the important question could be – is the `ProductsDataSource` designed well for our use case? 

> What are the risks of affecting the store-management section of the app (eg. introducing new bugs, etc)? 

An optimistic assumption is that we won't need to modify `ProductListHandler` implementation at all, which saves us development effort now. If we have to modify it in the future though, I'd consider forking/copying this class and adjusting it for the POS-specific purpose, or implementing from scratch. That said the store management app will not be affected at all.

> Does this decision make the store-management section more complex?

No

> Does it increase maintainability (eg. more testing is needed, code is more difficult to reason about, ...)?

Yes, any change to `ProductListHandler` will be reflected in both POS and store management modes. It will require additional awareness. If we want to avoid this, we can consider forking/copying this class.

> What are reasons for change to the ProductListHandler - is it dependant on the use-case (POS vs store-management) or is it a generic component?

My POV is that ProductListHandler is a generic component, mature, and with a narrow scope of responsibilities – a helper class for handling product pagination and search modes.

Let me know what you think, @malinajirka, @kidinov, @backwardstruck.

kidinov replied:
> I think the important question could be – is the ProductsDataSource designed well for our use case?

To me this is the main question. I personally haven't worked much with `ProductListHandler`, `ProductSelectorRepository` and `WCProductStore`. Is this any good? Idon't recall any complains that products are missing or not loading or something like that, so maybe a good sign. And also there is a good sign it doesn't do 10 network requests when products screen is open in the current app =)

API there is a bit weird though`loadFromCacheAndFetch` and `loadMore` returns `Result<Unit>` which we don't use at the moment

samiuelson replied:
Good point about unused returns – we don't handle errors in this PR. I think we should add some sort of `Result` return types to `loadProducts()` and `loadMore()` functions and handle errors in the VM (I added a [task](https://github.com/orgs/woocommerce/projects/247/views/3?pane=issue&itemId=63720240) for it).

Some other considerations related to interface design and DI:
* Do we want to support swipe to refresh
* Do we need a possibility to invalidate the data
* Do we want `ProductsDataSource` to be a singleton? This works as a naive in-memory data cache (optimizes perfomance), but also prevents from reloading products (risk of data being stale; also a problem in the app currently)

kidinov replied:
> Do we want to support swipe to refresh

I guess this is more design question, but I think yes we should

> Do we need a possibility to invalidate the data

I'd invalidate that on swipe to refresh gesture after success comes from the backend

> Do we want ProductsDataSource to be a singleton? This works as a naive in-memory data cache (optimizes perfomance), but also prevents from reloading products (risk of data being stale; also a problem in the app currently)

Won't see keep what we show in memory in VM anyway?

malinajirka replied:
> An optimistic assumption is that we won't need to modify ProductListHandler implementation at all, which saves us development effort now. If we have to modify it in the future though, I'd consider forking/copying this class and adjusting it for the POS-specific purpose, or implementing from scratch. That said the store management app will not be affected at all.

I think one more factor which might play a role here is that we might not be the developers who will need to modify the code in the future. Will the other devs realize, they should create a copy/replace/refactor the component instead of branching POS vs not-POS and therefore coupling these two modes?

Having said all that, I'm personally also inclined towards re-using it, since as you said `My POV is that ProductListHandler is a generic component, mature, and with a narrow scope of responsibilities – a helper class for handling product pagination and search modes.`. But in general, I think our approach to re-using needs to consider future implications and whether we are setting future-us and other-devs for a success or a failure.

samiuelson committed:
Merge branch 'trunk' into 11516-woo-pos-core-home-screen-ui--product-grid-cart-top-app-bar

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/cart/products/ProductSelector.kt:
```diff
@@ -0,0 +1,105 @@
+        }
+    }
+
+    LaunchedEffect(loadMore) {
```
❓ As you know I'm not that familiar with Compose. However, this feels a bit off to me from testability POV. Have you considered creating the flow in the VM and just collecting it here instead of building it here? Or is that a common practice?

samiuelson replied:
I can see that the naming is unfortunate here. The logic behind this code is to observe when the last element has been displayed on the screen to notify the VM about this. It's the VM's responsibility to consume this "user input" event – call the repository, update the ViewState model, etc. @malinajirka, do you think renaming `onLoadMore` -> `onEndOfProductsGridReached` like that: dd36fbe1e304240803a80f69420c81833057e0ea would make it less confusing or are you seeing another problem here that I'm missing?

malinajirka replied:
Ohh I see, I totally missed that. 

> do you think renaming onLoadMore -> onEndOfProductsGridReached like that:

Yes, I think this solves it :). Thanks!

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/cart/WooPosCartScreen.kt:
```diff
@@ -1,48 +1,120 @@
+@SuppressLint("UnusedMaterialScaffoldPaddingParameter")
 @Composable
-private fun WooPosCartScreen(onButtonClicked: () -> Unit) {
+private fun WooPosCartScreen(
```
🤔 I'm wondering if we could build the cart section and the product selector more independently. It seems we are assuming they'll always be on the same screen. Or perhaps the part which makes this a bit confusing for me is that `onLoadMore` feels irrelevant to the `cart` section but still is one of the parameters to WooPos**Cart**Screen. 

P.S. I realize this is just a skeleton, but 'onButtonClicked' feels like a very generic name. Could we perhaps rename it to something more specific to make the code easier to reason about?

kidinov replied:
I'd also reconsider packaging here and naming. Maybe call the screen with scaffold `CartBuildScreen` (and package as well) and inside of it can have 2 packages: `cart` and `productselector`?

samiuelson replied:
I split the cart and product selector into separate packages (4c7d9c8a78038a4e00b8dc16ef95023c4bc71a43). They are now parts of the Home screen and have separate VMs.

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/cart/products/ProductSelector.kt:
```diff
@@ -0,0 +1,105 @@
+    onLoadMore: () -> Unit,
+) {
+    ConstraintLayout(
+        modifier = Modifier.fillMaxWidth(0.7f)
```
💡 This is probably premature since we don't have final designs. However, in general I find it less error prone to define `Width` on the narrow component and `fillWidth` on the wide component in a two pane layout. Wdyt?

samiuelson replied:
Considering the designs are not stable yet, I'll leave it out for now.

malinajirka reviewed (approved):
Thanks @samiuelson! I've done a quick pass and left some questions. Please bear in mind that I'm not that familiar with Compose and not at all familiar with Compose navigation, so my comments might not make sense.

Overall, I didn't spot anything which would be blocking the merge. I'm approving it and leaving the merge to you or others.

samiuelson committed:
Rename function: onLoadMode 👉 onEndOfProductsGridReached

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/cart/products/ProductSelector.kt:
```diff
@@ -0,0 +1,105 @@
+        val state = productsState.collectAsState()
+        val gridState = rememberLazyGridState()
+        LazyVerticalGrid(
+            columns = GridCells.Adaptive(minSize = 128.dp),
```
np: don't we use direct improts for this line that?

```
horizontalArrangement = spacedBy(4.dp),
verticalArrangement = spacedBy(4.dp),
```

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/cart/products/ProductSelectorViewModel.kt:
```diff
@@ -0,0 +1,50 @@
+    savedStateHandle: SavedStateHandle,
+) : ScopedViewModel(savedStateHandle) {
+
+    val viewState: StateFlow<ViewState> = productsDataSource.products.map { products ->
```
Every time we go to the next screen and come back, we append a new page of the same products to the list. We probably initially did not do enough. Not sure if it's something that makes sense to fix at the current stage though

https://github.com/woocommerce/woocommerce-android/assets/4923871/87e56b16-13d5-488d-ba8a-ddfaab0ba997

samiuelson replied:
I don't see any duplicates. It looks like the next page of products gets loaded after coming back from Checkout. I'm not sure why it didn't load before opening the Checkout screen though. I logged a task [here](https://github.com/orgs/woocommerce/projects/247/views/3?pane=issue&itemId=63720240).

kidinov replied:
@samiuelson my bad, those are not duplicates but the next page indeed

samiuelson replied:
The problem occurs when the page size is so small all of its items are visible on the screen when rendered (no scroll is needed). I'll improve it in the next iterations.

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/cart/products/ProductsDataSource.kt:
```diff
@@ -0,0 +1,10 @@
+import com.woocommerce.android.model.Product
+import kotlinx.coroutines.flow.Flow
+
+interface ProductsDataSource {
```
I frankly don't see why we need an interface in this case. `ProductsDataSourceImpl` defines an interface with it's public methods anyway

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/cart/products/ProductSelectorViewModel.kt:
```diff
@@ -0,0 +1,50 @@
+        }
+    }
+
+    data class ViewState(
```
np: I'd move those to a separate class to avoid VMs with 2000 lines of code as we have now and to define solid files structure from the ground up

malinajirka replied:
+1 I think anything that can be extracted from a VM should live in an encapsulated file.

kidinov reviewed (commented):
LGTM. I left a few suggestion, please take a look

samiuelson committed:
Rename param

samiuelson committed:
Reorganise packages

samiuelson committed:
Optimise imports

samiuelson committed:
Extract view state class to separate file

samiuelson committed:
Extract view state class to separate file

samiuelson committed:
Merge remote-tracking branch 'origin/11516-woo-pos-core-home-screen-ui--product-grid-cart-top-app-bar' into 11516-woo-pos-core-home-screen-ui--product-grid-cart-top-app-bar

samiuelson committed:
Fix detekt's complaints

samiuelson committed:
Clean up code

samiuelson committed:
Simplify UI layouts

samiuelson commented:
I'm converting it to draft PR because we want to focus first on implementing the whole navigation flow rather than individual screens. This PR can be reused or serve as inspiration for implementing the product selector pane though.

malinajirka commented:
>  we want to focus first on implementing the whole navigation flow rather than individual screens. 

@samiuelson Could you please share an additional context about this direction - a link to a related conversation is fine. Thanks!

samiuelson committed:
Merge branch `11516-woo-pos-core-home-screen-ui--product-grid-cart-top-app-bar`

samiuelson committed:
Update grid UI

samiuelson committed:
Update grid UI

samiuelson committed:
Adjust colors for dark/light theme

samiuelson commented:
> 
> @samiuelson Could you please share an additional context about this direction - a link to a related conversation is fine. Thanks!

@malinajirka, p1717001202981269/1716809008.635649-slack-C070SJRA8DP During the POS sync meeting we analyzed how the UX concepts are evolving (i1 vs i2) and realized that until the UX flow is final we won't know exactly how to share state between screens/panes. That's why we asked to speed up the final decision on the UX last week. I'm resuming the work on this PR as we have clarity on the navigation.

samiuelson committed:
Decouple product selector from cart

samiuelson committed:
Merge branch 'trunk' into 11516-woo-pos-core-home-screen-ui--product-grid-cart-top-app-bar

samiuelson commented:
👋🏼 @malinajirka, @kidinov I've addressed your comments, and updated the branch with the `11644-woo-pos-implement-skeleton-of-the-i3-of-the-wireframes-2` branch containing the static bottom toolbar. Could you give it another look?
 
https://github.com/woocommerce/woocommerce-android/assets/4527432/ce3ddfd2-7ee4-44ec-8a38-1970c88b614f

In the next iteration, I will focus on syncing item selection state between Cart and Product Selector and creating the Order.

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/WooPosHomeNavigation.kt:
```diff
@@ -0,0 +1,22 @@
+
+internal const val HOME_ROUTE = "home"
+
+internal fun NavGraphBuilder.homeScreen(onCheckoutClick: () -> Unit) {
```
I changed "Cart" to "Home" as the cart is only one of the panes on the home screen.

malinajirka reviewed (approved):
@samiuelson Thank you for working on this. I've reviewed the PR and re-tested it locally - I think it's a great prototype we can iterate on. I'd merge it, but it seems the CI is failing - feel free to merge it after that's fixed.

samiuelson committed:
Fix lint

samiuelson committed:
Fill product selector's height
